keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+
+Proposed logger implementation are not designed for extensibility. Any change of behavior would require a change in code by creating PR. Looking at kubernetes history we know that this will not be maintainable. Eventually kubernetes adapted dedicated interfaces like CSI, CRI, CNI. We think that similar think could happen for logging in the future, but it doesn’t need to be introduced immediately. A good point for starting this discussion would be when this effort would be GAed. Then we could propose another logging API implementation.
+
+#### Huge increase of log volume
+
+This effort should result in reducing costs of log storage and analysis in production setups. Introducing structured logs will allow for creating better indices and reducing the size of data needed for analysis. From log throughput perspective there will be an increase of log size causing a pressure of log ingestion (disk, logging agents, logging API). To reduce the potential impact we will consider adding additional configuration flags allowing user to specify which metadata should be dropped
+
+## Design Details
+
+### Mapping between logging methods
+
+Following table provides a mapping between klog and logr methods to be used during the automated migration.
+
+| **klog logger** | **logr logger**
+| - |-  |
+`klog.V(level).Info*(format/args[0], args)` | `<module>.Logger.V(level+2).Info(format/args[0], “args”, args)`
+`klog.Info*(format/args[0], args)` | `<module>.Logger.V(2).Info(format/args[0], “args”, args)`
+`klog.Warning*(format/args[0], args)` | `<module>.Logger.V(1).Info(format/args[0], “args”, args)`
+`klog.Error*(format/args[0], args)` | `<module>.Logger.Error(error(format/args[0]), format/args[0], “args”, args)` or after manual verification that this is not a proper error log: `<module>.Logger.Info(format/args[0], “error”, error(format/args[0]), “args”, args)`
+`klog.Fatal*(format/args[0], args)` | `<module>.Logger.Error(error(format/args[0]), format/args[0], “args”, args, “stacks”, k8slog.AllStacks()) os.Exit(255)`
+`klog.Exit*(format/args[0], args)` | `<module>.Logger.Error(error(format/args[0]), format/args[0], “args”, args) os.Exit(1)`
+
+### Test Plan
+
+TBD
+
+### Graduation Criteria
+
+#### Alpha
+
+* Introduce a flag allowing to switch to new log format
+* Mechanism for defining standard key fields is implemented
+* `klog` is deprecated and should not be used for new logs
+* Automatic stage of migration is finished 
+
+#### Beta
+
+* Structured logs were manually verified 
+* New Logging guide for K8s is available
+
+#### GA
+ 
+* Structured logs are the default
+* Instructions on migrating other kubernetes repositories
+
+#### Deprecating old logging format
+
+* After 2 k8s versions old logging format removed 
+
+### Upgrade / Downgrade Strategy
+
+Double writing during migration will ensure identical logs without any configuration changes. Switching to new format will require passing a flag to each component or waiting for full removal of klog.
```

Comment:
What do you mean by "double-writing"? Depending on the exact meaning, it may be a significant performance hit.

Reply:
"double-writing" in this context means using two different APIs for logging (klog and logr) in the same time. Actual logs produced by each component will not be doubled and will use just one format.

Reply:
I'm a little confused about this as the proposal starts with "Double writing using both klog and logr APIs for new log entries should not be needed."